curator-recipes/src/main/java/org/apache/curator/framework/recipes/leader/LeaderLatch.java

```diff
@@ -667,9 +670,9 @@ protected void handleStateChange(ConnectionState newState)
             {
                 try
                 {
-                    if ( client.getConnectionStateErrorPolicy().isErrorState(ConnectionState.SUSPENDED) || !hasLeadership.get() )
```

Comment:
I'm wondering whether we should remove this `if` entirely: Let's assume the current instance doesn't have the leadership but watches the child node referring to the current leader. If the connection was temporarily suspended it could be that the actual leader lost its leadership in the meantime. In that case, we want the current instance to check the leadership. :thinking:

Reply:
But thinking about it once more, this might not be an issue because the current LeaderLatch would have a watcher on the current leader's node which would have been triggered when the leader's child would have been deleted. :thinking:

Reply:
It's for backward compatibility that once we treat ConnectionLoss as ErrorState. I agree with that since we don't give up the leadership (`reset`) actively, we can remove the `if` condition (but still keep `getChildren`, of course).

However, it can be a follow-up improvement that doesn't affect the correctness. You're welcome to make the patch.

Reply:
If you take a look at FLINK-10052, the final solution is using a SessionErrorPolicy that will skip this `if` block. Since a ConnectionLoss may be only network unstable instead of the node lost its leadership (the ephemeral node). Before this patch it's `reset` to be called and actively give up the leadership, it will cause reelection, increase ZK workload and cause further inconsistency.

The thorough solution should be something like I proposed in https://github.com/apache/flink/pull/9878, but I failed to contribute it to the upstream (FLINK-10052 takes more than 2 years to be merged. It's not a good experience to me, lol). We run with the solution in Tencent for years and it works well :)

Reply:
I'm reverting this change because it's a OR logic but I misread it as an AND logic in the first place.

Reply:
... while I think we can remove the whole if condition (keep the `getChildren` call. You can give it a try.